models/service.go

```diff
@@ -17,46 +17,58 @@ type ServiceList struct {
 }
 
 type Service struct {
-	Name         string              `json:"name"`
-	Namespace    Namespace           `json:"namespace"`
-	Labels       map[string]string   `json:"labels"`
-	Type         string              `json:"type"`
-	Ip           string              `json:"ip"`
-	Ports        Ports               `json:"ports"`
-	Endpoints    Endpoints           `json:"endpoints"`
-	Pods         Pods                `json:"pods"`
-	RouteRules   RouteRules          `json:"route_rules"`
-	Dependencies map[string][]string `json:"dependencies"`
+	istioClient      kubernetes.IstioClientInterface
```

Comment:
I am a bit confused by the "model" responsibilities here. In general models are much more straightforward structures that just hold data and eventually perform some manipulation / conversion (such as the "Parse" functions below), but having the model depending on client backends in a bit surprising for me (maybe I'm too much conventional :-) ). If we could separate as much as possible the pure model-kind functions and the clients calls it would make unit-testing and mocking easier.

Anyway since it would require a non trivial refactoring maybe we can ignore this remark for now.

Reply:
Yes, this is also weird to me.
The discussion for this model was to separate two domains:
- the K8S, Istio and OS domain which model is huge
- the SWS model used to support SWS REST endpoints (so, then a small subset of the first domain).

So, I know that while coding there is a chance that to mix those, but I think we should maintain those two layers, and maintain that philosophy, if not, in two months people won't know why each model was created.

Reply:
So an option for refactoring could be:

- Having the handler (not the model) calling clients to get the low-level model
- The handler calls SWS model setters to fill structures using the low-level model
- We can easily build unit-test without any mock injection to test model conversion (from low-level to SWS model)

Reply:
At the beginning I was thinking about an Active Record paradigm where models are the ones dealing with the data storage/stream. That is also why I used an interface for the clients (maybe names aren't the best). If those clients change, they would have to implement same methods.

Another reason I decided to go that way was because I didn't want to overload handlers and avoid spaghetti code. Controllers or handlers might easily get really fat due to too many responsibilities.

I'll go with @jotak approach and ping you back.